### PR TITLE
P2P add PeerServer class - Closes  #4638 & #4789

### LINKED ISSUE #4638: P2P move PeerServer logic to it is own class
### Description
- Currently we are initiating and handling the PeerServer inside the P2P main class, it should be a logically separated class.

### Motivation
- The codebase will be more readable and allow further improvements in the PeerServer.

### Acceptance Criteria
- Create individual PeerServer class and keep all the existing features.

### PR DESCRIPTION
**Pull request by valamidev**
### What was the problem?
- Currently we are initiating and handling the PeerServer inside the P2P main class, it should be a logically separated class.

### How did I solve it?
- Create a new Class called PeerServer which handle all the inbound connections.

### How to manually test it?

`npm run test
`
### Review checklist

- [x] The PR resolves #4638 & #4789
- [x] All new code is covered with unit tests
- [x] Relevant integration / functional tests are added
- [x] Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
- [x] Documentation has been added/updated

### CODE REVIEW

File: elements/lisk-p2p/src/p2p.ts [hunk 1]
```diff
@@ -268,19 +242,11 @@ export class P2P extends EventEmitter {
 		this._config = config;
 		this._isActive = false;
 		this._hasConnected = false;
+		this._peerServer = undefined;
```
Comment (ishantiw) [inline elements/lisk-p2p/src/p2p.ts]: we can use `?` for  `private _peerServer?: PeerServer;` then we don't have to initialize it to `undefined`

File: elements/lisk-p2p/src/p2p.ts [hunk 2]
```diff
@@ -358,6 +324,18 @@ export class P2P extends EventEmitter {
 			this.emit(EVENT_CLOSE_INBOUND, closePacket);
 		};
 
+		this._handleInboundPeerConnect = (
+			incomingPeerConnection: IncomingPeerConnection,
+		) => {
+			this._peerPool.addInboundPeer(
```
Comment (ishantiw) [inline elements/lisk-p2p/src/p2p.ts]: This function throws an error when it already exists in peer pool, we should wrap it around `try/catch` and emit relevant events

File: elements/lisk-p2p/src/p2p.ts [hunk 3]
```diff
@@ -936,7 +647,33 @@ export class P2P extends EventEmitter {
 			throw new Error('Node cannot start because it is already active.');
 		}
 
-		await this._startPeerServer();
+		if (this._config.maxOutboundConnections !== 0) {
```
Comment (ishantiw) [inline elements/lisk-p2p/src/p2p.ts]: shouldn't it be `this._config.maxInboundConnections`?

File: elements/lisk-p2p/src/p2p.ts [hunk 4]
```diff
@@ -997,5 +738,9 @@ export class P2P extends EventEmitter {
 		peerPool.on(EVENT_INBOUND_SOCKET_ERROR, this._handleInboundSocketError);
 		peerPool.on(EVENT_BAN_PEER, this._handleBanPeer);
 	}
+
+	private _bindHandlersToPeerServer(peerServer: PeerServer): void {
+		peerServer.on(EVENT_NEW_INBOUND_PEER, this._handleInboundPeerConnect);
```
Comment (ishantiw) [inline elements/lisk-p2p/src/p2p.ts]: I think we should also bind `EVENT_FAILED_TO_ADD_INBOUND_PEER` that `peerServer` is emitting as well

Comment (valamidev) [reply elements/lisk-p2p/src/p2p.ts]: Seems `EVENT_FAILED_TO_ADD_INBOUND_PEER` could only hold new information if the PeerPool.addInboundPeer() is failing due duplication. 
Any other case EVENT_INBOUND_SOCKET_ERROR is holding all the information.

File: elements/lisk-p2p/test/unit/p2p.ts [hunk 5]
```diff
@@ -75,24 +82,59 @@ describe('p2p', () => {
...
 		});
 
 		it('should reject at multiple stop attempt', async () => {
 			await P2PNode.stop();
-
 			expect(P2PNode.stop()).rejects.toThrow();
 		});
 	});
+
+	describe('#PeerServer', () => {
```
Comment (ishantiw) [inline elements/lisk-p2p/test/unit/p2p.ts]: it looks unusual that we are having a unit test based on real sockets and still mocking some functions of PeerPool. Need to think about making it consistent

File: elements/lisk-p2p/src/p2p.ts [hunk 6]
```diff
@@ -997,5 +774,18 @@ export class P2P extends EventEmitter {
 		peerPool.on(EVENT_INBOUND_SOCKET_ERROR, this._handleInboundSocketError);
 		peerPool.on(EVENT_BAN_PEER, this._handleBanPeer);
 	}
+
+	private _bindHandlersToPeerServer(peerServer: PeerServer): void {
```
[the remark this answers is not in the document]
Comment (valamidev) [reply elements/lisk-p2p/src/p2p.ts]: Yes, and we can unbind the PeerPool listeners as well.

File: elements/lisk-p2p/src/peer_server/peer_server.ts [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+		if (!socket.request.url) {
+			this._disconnectSocketDueToFailedHandshake(
+				socket,
+				INVALID_CONNECTION_URL_CODE,
+				INVALID_CONNECTION_URL_REASON,
+			);
+
+			return;
+		}
+		const queryObject = url.parse(socket.request.url, true).query;
```
Comment (ManuGowda) [inline elements/lisk-p2p/src/peer_server/peer_server.ts]: `const { query } = url.parse(socket.request.url, true).query;`?

File: elements/lisk-p2p/src/peer_server/peer_server.ts [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+
+		if (queryObject.nonce === this._nodeInfo.nonce) {
```
Comment (ManuGowda) [inline elements/lisk-p2p/src/peer_server/peer_server.ts]: check for the query? `if (queryObject && query.nonce === this._nodeInfo.nonce) {`

File: elements/lisk-p2p/src/peer_server/peer_server.ts [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+			this._peerBook.removePeer({
+				peerId: constructPeerId(socket.remoteAddress, selfWSPort),
+				ipAddress: socket.remoteAddress,
+				wsPort: selfWSPort,
+			});
+
+			return;
+		}
+
+		if (
```
Comment (ManuGowda) [inline elements/lisk-p2p/src/peer_server/peer_server.ts]: shouldn't this if check come before the previous one?

File: elements/lisk-p2p/src/peer_server/peer_server.ts [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+			new PeerInboundHandshakeError(
+				closeReason,
+				statusCode,
+				socket.remoteAddress,
+				socket.request.url,
+			),
+		);
+	}
+
+	private _handleIncomingConnection(socket: SCServerSocket): void {
```
Comment (ManuGowda) [inline elements/lisk-p2p/src/peer_server/peer_server.ts]: It's hard to follow the this function as its huge, around 167 lines of code. can we split the functionality into following:
- `_validateQueryObject`
- `_preparePeerInfoForConnection`
- `_validatePeerInfo`
- `_makeHandshake`

I suggested the function names according to the code flow, you can use names according to your view as well. 
what you think @ishantiw ?

File: elements/lisk-p2p/src/peer_server/peer_server.ts [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+	private _handleIncomingPayload(ws: any, _req: any): void {
+		ws.on('message', (message: any) => {
+			// Pong message
+			if (message === '#2') {
+				return;
+			}
+
+			const MAX_EVENT_NAME_LENGTH = 128;
+
+			const peerIpAddress = ws._socket._peername.address;
```
Comment (ManuGowda) [inline elements/lisk-p2p/src/peer_server/peer_server.ts]:
```
const { address, port } = ws._socket._peername;
const peerId = `${address}:${port}`;
```

File: elements/lisk-p2p/src/peer_server/peer_server.ts [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+					'#unsubscribe',
+					'#publish',
+				]);
+
+				if (
+					(parsed.event && typeof parsed.event !== 'string') ||
+					invalidEvents.has(parsed.event) ||
+					parsed.event.length > MAX_EVENT_NAME_LENGTH
+				) {
+					throw new Error('Received invalid payload');
```
Comment (ManuGowda) [inline elements/lisk-p2p/src/peer_server/peer_server.ts]: `throw new Error('Received invalid payload', parsed);` this way we could know the context better while debugging

File: elements/lisk-p2p/src/peer_server/peer_server.ts [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+				}
+
+				if (parsed.event === '#disconnect') {
+					const count =
+						(this._invalidMessageCounter.get(peerIpAddress) || 0) + 1;
+					this._invalidMessageCounter.set(peerIpAddress, count);
+
+					if (count > DEFAULT_CONTROL_MESSAGE_LIMIT) {
+						throw new Error('Received invalid payload');
```
Comment (ManuGowda) [inline elements/lisk-p2p/src/peer_server/peer_server.ts]:
```
throw new Error(`Exhausted disconnect event DEFAULT_CONTROL_MESSAGE_LIMIT: ${DEFAULT_CONTROL_MESSAGE_LIMIT}`);
```

File: elements/lisk-p2p/src/peer_server/peer_server.ts (hunk 10 above)
Comment (ishantiw) [reply elements/lisk-p2p/src/peer_server/peer_server.ts]: Sounds good 👍

File: elements/lisk-p2p/src/p2p.ts [hunk 14]
```diff
@@ -959,7 +727,14 @@ export class P2P extends EventEmitter {
...
 		this._peerPool.removeAllPeers();
-		await this._stopPeerServer();
+		this._peerBook.cleanUpTimers();
+
+		if (this._peerServer) {
+			await this._peerServer.stop();
+			this._removeListeners(this._peerServer);
+		}
+
+		this._removeListeners(this._peerPool);
```
[the remark this answers is not in the document]
Comment (valamidev) [reply elements/lisk-p2p/src/p2p.ts]: PeerServer is stopped, but the peerServer object is still existing and P2P listener are binded to it. It is more clean if we remove these bindings. Same for peerPool.

File: elements/lisk-p2p/src/peer_server/peer_server.ts (hunk 9 above)
Comment (ishantiw) [reply elements/lisk-p2p/src/peer_server/peer_server.ts]: I think its good for an early return, as soon as we detect connect to itself we return
